PEP 812: Imaginary type and IEC 60559-compatible complex arithmetic (#4681)
python/peps: peps/pep-0812.rst

``~>`` denotes type coersion and ``~=`` is substitution of ``complex(0, y)``
for ``yj``):

.. code:: python

   2.0 * (inf+3j) ~> complex(2, 0) * complex(inf, 3)
   == complex(2.0*inf - 0.0*3.0, 2.0*3.0 + 0.0*inf)
   == complex(inf, nan)

.. code:: python

   2j * (inf+3j) ~= complex(0, 2) * complex(inf, 3)
   == complex(0.0*inf - 2.0*3.0, 2.0*inf + 0.0*3.0)
   == complex(nan, inf)

.. code:: python

   2j * complex(-0.0, 3) ~= complex(0, 2) * complex(-0.0, 3)
   == complex(-0.0*0.0 - 2.0*3.0, -2.0*0.0 + 0.0*3.0)
   == complex(-6.0, 0.0)

Users with some mathematical background instead would expect, that ``x+yj``
notation in the Python is equal to ``complex(x, y)`` and is just the usual
rectangular form for a complex number with components ``x`` (real) and ``y``
(imaginary) and that above examples will work, following rules of elementary
algebra (with assumption that ``1j**2`` is ``-1``):

.. code:: python

   2.0 * (inf+3j) == complex(2.0*inf, 2.0*3.0) == complex(inf, 6)
   2j * (inf+3j) == complex(-2.0*3.0, 2.0*inf) == complex(-6, inf)
   2j * complex(-0.0, 3) == complex(-2.0*3.0, -2.0*0.0) == complex(-6, -0.0)

> Same affects addition (or subtraction):

Reviewer: ```suggestion
The same affects addition (or subtraction):
```

arithmetic implemented in the Ada language [5]_.  Some mathematical libraries
(like the GNU GSL [6]_) or the GNU MPC [7]_) have special routines to implement
mixed-mode arithmetic for complex numbers, i.e. when one operand either
pure-real or pure-imaginary.  As a side effect, this also introduce some
performance boost for operations with mixed types (e.g. multiplication of
complex and real numbers costs only two real multiplications, not four).
Though it's more important, that in the IEC floating-point arithmetic results
here are uniquely determined by usual mathematical formulae.

For a first step, :cpython-pr:`124829` added in the CPython 3.14 mixed-mode
rules for complex arithmetic, combining real and complex operands.  So, some
examples from above now are working correctly:

.. code:: pycon

   >>> from cmath import inf
   >>> 2*(inf+3j)
   (inf+6j)
   >>> 2-0j
   (2-0j)

Unfortunately, this is only a half-way solution.  To fix the rest of examples
we need a separate type for pure-imaginary complex numbers.

Rationale
=========

Lets collect here some arguments (with possible answers) against adoption of

> newer approach to complex arithmetic in Python.

Reviewer: ```suggestion
Let's collect here some arguments (with possible answers) against adoption of
a newer approach to complex arithmetic in Python.
```

Special cases are rare
----------------------

That's not true, as special numbers coming not just from input, but also
*during* computations (e.g. with underflow or overflow).  Robust software must
account for them and currently this usually require to reinvent complex
arithmetic in application, i.e. use the :external+py3.14:class:`complex` type
just as a bag for it's components.  The only known cure for this is presented
by the PEP:

    Generally, mixed-mode arithmetic combining real and complex variables
    should be performed directly, not by first coercing the real to complex,
    lest the sign of zero be rendered uninformative; the same goes for
    combinations of pure imaginary quantities with complex variables.

    -- Kahan, W: `Branch cuts for complex elementary functions
    <https://people.freebsd.org/~das/kahan86branch.pdf>`_.

The complex facility should be simple
-------------------------------------

Simplicity is a goal, but most importantly for the end user, not implementators
of the complex arithmetic.

Upcoming C2y standard abandon [8]_ the imaginary type

> -----------------------------------------------------

Reviewer: Also this footnote ref breaks the header:

<img width="781" height="116" alt="image" src="https://github.com/user-attachments/assets/aa3c55ae-718d-4078-945e-8d4e80e0c322" />

https://pep-previews--4681.org.readthedocs.build/pep-0812/#id8

Please can you move it into the prose?

Parsing strings with the integer "negative zero" in real part (i.e. ``"-0+1j"``
or ``"(-0+1j)"``) will be deprecated in the :external+py3.14:class:`complex`
constructor.

The :mod:`marshal` module will be adjusted to support new type.

New C-API
---------

.. c:type:: PyImaginaryObject

   This subtype of :c:type:`PyComplexObject` represents a Python imaginary
   number object.

.. c:var:: PyTypeObject PyImaginary_Type

   This instance of :c:type:`PyTypeObject` represents purely imaginary numbers,
   the Python complex number type *without* real component.

.. c:function:: int PyImaginary_Check(PyObject *p)

   Return true if its argument is a :c:type:`PyImaginaryObject` or a subtype of
   :c:type:`PyImaginaryObject`.  This function always succeeds.

.. c:function:: int PyImaginary_CheckExact(PyObject *p)

   Return true if its argument is a :c:type:`PyImaginaryObject`, but not a
   subtype of :c:type:`PyImaginaryObject`.  This function always succeeds.

.. c:function:: PyObject* PyImaginary_FromDouble(double imag)

   Return a new :c:type:`PyImaginaryObject` object with *imag* imaginary
   component.  Return ``NULL`` with an exception set on error.

>    Imaginary component value of a new object could be taken with

Reviewer: ```suggestion
   The imaginary component value of a new object could be taken with
```

   :c:func:`PyComplex_ImagAsDouble`.

In conformance with recent C-API group `decision
<https://github.com/capi-workgroup/decisions/issues/56>`__ we don't offer API
to do arithmetic on low-level representation of complex numbers in CPython.
Instead, it's expected that C-API users either will use  `PyNumber_*
<https://docs.python.org/3/c-api/number.html>`__ API or will export numbers
from Python objects and do arithmetic with some external library (like the GNU
GSL), then import back.

> It's Not Magic

Reviewer: ```suggestion
It's mot magic
```

   (-1.5707963267948966+0.5493061443340549j)

Though, we should mention that floating-point arithmetic is not a replacement
for ``limit()`` facilities of computer algebra systems.  Using same identity
near real line will show wrong results:

.. code:: pycon

   >>> z = 2+0j
   >>> cmath.atan(z)
   (1.1071487177940904+0j)
   >>> atan(z)
   (1.1071487177940904+0j)
   >>> z = 2-0j
   >>> cmath.atan(z)
   (1.1071487177940904-0j)
   >>> atan(z)
   (1.1071487177940904+0j)

Of course, similar happens already for real floating-point arithmetic:

.. code:: pycon

   >>> f = lambda x: (1 + x)/(1 - x) - 1
   >>> f(1e-15)
   2.220446049250313e-15
   >>> f(0.0)
   0.0
   >>> f(-0.0)
   0.0
   >>> f(-1e-15)
   -2.1094237467877974e-15

Applications must carefully choose expressions from equivalent forms.

Backwards Compatibility
=======================

In one sense, this PEP should have relatively low impact for end users.

Indeed, no new syntax introduced.  Results for complex arithmetic will be
different, if computation trigger some corner cases, where before either
meaningless values were obtained (``nan``'s) or wrong zero signs.  In the

> later case, results will be indistinguishable for equality (``==``) testing.

Reviewer: ```suggestion
Indeed, no new syntax is introduced.  The results for complex arithmetic will be
different if computation triggers some corner cases, where before either
meaningless values were obtained (``nan``\ s) or wrong zero signs.  In the
latter case, the results will be indistinguishable from equality (``==``) testing.
```

"some complex numbers", to produce exceptions.  Running the `mpmath
<https://github.com/mpmath/mpmath>`_ (develompment version) test suite shows
only two test failures.  The `NumPy <https://github.com/numpy/numpy>`_ (v2.3.4)
has ~16 broken tests.

How to Teach This
=================

While internaly complex arithmetic will be more complicated (but not too much,
see `Reference Implementation <Reference Implementation_>`_), its semantics
will be more close to usual mathematical notation in textbooks on complex
analysis, much less place for confusion of newcomers.  Roughly speaking, it
will be the floating-point arithmetic, augmented by the special algebraical
symbol ``1j``, which square is ``-1.0``.

Reference Implementation
========================

A draft implementation is available in a
https://github.com/skirpichev/cpython/pull/1

Rejected Ideas
==============

We might try to implement complex arithmetic, that will treat specially ---
just like instances of imaginary type proposed --- complex numbers of the form
``complex(0.0, y)``.  But such proposal alter arithmetic rules on the set of
complex numbers itself, in particular ``complex(a, b) + complex(c, d)`` will

> not be exactly equal to ``complex(a + c, b + d)`` anymore.

Reviewer: ```suggestion
We might try to implement complex arithmetic that will specially treat complex
numbers of the form ``complex(0.0, y)`` --- just like instances of imaginary type
proposed.  But such a proposal alters the arithmetic rules on the set of
complex numbers itself, in particular ``complex(a, b) + complex(c, d)`` will
not be exactly equal to ``complex(a + c, b + d)`` anymore.
```

The set of imaginary numbers with a special treatment in complex arithmetic
might implemented differently, as a distinct form for the complex type
constructor, say ``complex(imag=y, pure=True)``.  However, experiments show
that such implementation is more complicated internally and more hard to

> explain then a dedicated concept of imaginary numbers as a subtype of complex.

Reviewer: ```suggestion
The set of imaginary numbers with special treatment in complex arithmetic
might be implemented differently, as a distinct form for the complex type
constructor, say ``complex(imag=y, pure=True)``.  However, experiments show
that such an implementation is more complicated internally and harder to
explain then a dedicated concept of imaginary numbers as a subtype of complex.
```

Open Issues
===========

The PEP doesn't expose new subtype as a builtin, say ``imaginary``, but maybe
we should?  This looks redundant, as all imaginary values could be obtained by

> scaling imaginary unit, i.e. ``imaginary(x) == float(x)*1j``.

Reviewer: ```suggestion
The PEP doesn't expose a new subtype as a builtin, say ``imaginary``, but maybe
we should?  This looks redundant, as all imaginary values could be obtained by
scaling imaginary units, i.e. ``imaginary(x) == float(x)*1j``.
```

Author: This changes meaning.  "the imaginary unit" - will be correct, I think.

           indeed anything else.

           -- https://j3-fortran.org/doc/year/24/24-179.txt

.. [2] W. Kahan and J. W. Thomas.  `Augmenting a Programming Language with
   Complex Arithmetic
   <http://www2.eecs.berkeley.edu/Pubs/TechRpts/1992/CSD-92-667.pdf>`_.
   Technical Report UCB/CSD 91/667, Univ. of Calif. at Berkeley, December,
   1991.

.. [3] ISO/IEC 9899:1999, Annex G.  See `N1256 (final draft)
   <https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf>`_.
   https://open-std.org/, WG14. 2007.

.. [4] See `Rationale for C99
   <https://www.open-std.org/jtc1/sc22/wg14/www/C99RationaleV5.10.pdf>`_.
   https://open-std.org/, WG14, 2003 and `Issues Regarding Imaginary Types for
   C and C++
   <https://wiki.edg.com/pub/CFP/WebHome/imaginary%20types%20in%20C--Thomas-Coonen.pdf>`_,
   by Jim Thomas and Jerome T. Coonen, The Journal of C Language Translation,
   Volume 5, Number 3, March 1994.

.. [5] `Ada 2022 Reference Manual
   <http://ada-auth.org/standards/22rm/RM-Final.pdf>`_, Annex G.

.. [6] `GNU Scientific Library, Release 2.7
   <https://www.gnu.org/software/gsl/doc/latex/gsl-ref.pdf>`_, §5.5.

.. [7] `The GNU Multiple Precision Complex Library, Edition 1.3.1

>    <https://www.multiprecision.org/downloads/mpc-1.3.1.pdf>`_, §5.7.

Reviewer: We prefer to avoid using footnotes when there's not additional info in the footnote, so we can save our readers a click to get to the important info. I think 5-7 would be better as inline links in the prose.

Author: I don't think they are principally different from the rest.  Important info already included in the text.  Footnotes are for additional stuff and/or references.
